harness/determined/common/api/_session.py

```diff
@@ -1,27 +1,93 @@
-from typing import Any, Dict, Optional
+import abc
+import json as _json
+from typing import Any, Dict, Optional, Tuple, Union
 
 import requests
 import urllib3
 
+import determined as det
+from determined.common import requests as det_requests
 from determined.common import util
-from determined.common.api import authentication, certs, request
+from determined.common.api import authentication, certs, errors, request
 
 
-class Session:
-    def __init__(
-        self,
-        master: Optional[str],
-        user: Optional[str],
-        auth: Optional[authentication.Authentication],
-        cert: Optional[certs.Cert],
-        max_retries: Optional[urllib3.util.retry.Retry] = None,
-    ) -> None:
-        self._master = master or util.get_default_master_address()
-        self._user = user
-        self._auth = auth
-        self._cert = cert
-        self._max_retries = max_retries
+def _do_request(
+    method: str,
+    host: str,
+    path: str,
+    params: Optional[Dict[str, Any]] = None,
+    json: Any = None,
+    data: Optional[str] = None,
+    headers: Optional[Dict[str, str]] = None,
+    cert: Optional[certs.Cert] = None,
+    timeout: Optional[Union[Tuple, float]] = None,
+    stream: bool = False,
+    max_retries: Optional[urllib3.util.retry.Retry] = None,
+) -> requests.Response:
+    # Allow the json json to come pre-encoded, if we need custom encoding.
+    if json is not None and data is not None:
+        raise ValueError("json and data must not be provided together")
+
+    if json:
+        data = det.util.json_encode(json)
+
+    try:
+        r = det_requests.request(
```

Comment:
woah this is confusing, there's a whole other separate requests module... 

might be out of scope for this PR, but it'd be nice if we could consolidate that logic in here and get rid of it.

Reply:
not sure if there's a good reason we insist on using `requests` and later on end up adding underlying `urllib3` adapters anyway. we should just use `urllib3` here. if we do that, the whole "certs with different server host names" issue is much cleaner to solve:

```
pool = urllib3.HTTPSConnectionPool(
    host,
    assert_hostname=cert.name,
)
pool.request(...)
```